### PR TITLE
Fixing plural/singular issues in the po parser

### PR DESCRIPTION
**Pull request by burzum**
This will fix the case described here: https://github.com/cakephp/cakephp/issues/10730#issuecomment-349286149

### CODE REVIEW

File: tests/TestCase/I18n/Parser/PoFileParserTest.php [hunk 1]
```diff
@@ -65,39 +65,39 @@ public function testParse()
         $expected = [
             'Plural Rule 1' => [
                 '_context' => [
-                    '' => 'Plural Rule 1 (translated)',
-                ],
```
Comment (dereuromark) [inline tests/TestCase/I18n/Parser/PoFileParserTest.php]: You shouldnt revert the trailing commas IMO, just too much diff changes then that are none.

File: tests/TestCase/I18n/Parser/PoFileParserTest.php [hunk 2]
```diff
@@ -65,39 +65,39 @@ public function testParse()
...
                     'Another Context' => '%d = 1 (translated)'
                 ]
             ],
-            '%d = 0 or > 1' => [
+            'p:%d = 0 or > 1' => [
                 '_context' => [
                     'Another Context' => [
-                        0 => '%d = 1 (translated)',
-                        1 => '%d = 0 or > 1 (translated)'
+                        (int)0 => '%d = 1 (translated)',
```
Comment (dereuromark) [inline tests/TestCase/I18n/Parser/PoFileParserTest.php]: I dont think you need the (int) cast here on the keys.

File: src/I18n/Translator.php [hunk 3]
```diff
@@ -22,6 +22,8 @@
 class Translator extends BaseTranslator
 {
 
+    const PLURAL_PREFIX = 'p:';
```
Comment (markstory) [inline src/I18n/Translator.php]: Could we have one copy of the constant?

File: src/I18n/Translator.php [hunk 4]
```diff
@@ -33,7 +35,18 @@ class Translator extends BaseTranslator
      */
     public function translate($key, array $tokensValues = [])
     {
-        $message = $this->getMessage($key);
+        if (isset($tokensValues['_count'])) {
+            $message = $this->getMessage(static::PLURAL_PREFIX . $key);
+            if (!$message) {
+                $message = $this->getMessage($key);
+            }
```
Comment (markstory) [inline src/I18n/Translator.php]: We'll need this for backwards compatibilty and userland defined message loaders/parsers.

File: tests/TestCase/I18n/Parser/PoFileParserTest.php (hunk 1 above)
Comment (burzum) [reply tests/TestCase/I18n/Parser/PoFileParserTest.php]: Seriously? All the other lines haven't had one I think and AFAIK best practice by CS is to not have it there?

File: src/I18n/Translator.php (hunk 4 above)
Comment (burzum) [reply src/I18n/Translator.php]: What exactly do you mean? Do I need to change something else here?

File: tests/TestCase/I18n/Parser/PoFileParserTest.php (hunk 1 above)
Comment (dereuromark) [reply tests/TestCase/I18n/Parser/PoFileParserTest.php]: No best practice (what the sniffer should be enforcing) is to always have the multiline trailing commas, to keep diffs low exactly for the reasons shown in this PR: Too much unrelated noise around the few lines of actual change.

Comment (burzum) [reply tests/TestCase/I18n/Parser/PoFileParserTest.php]: They aren't used everywhere either here! https://github.com/cakephp/cakephp/blob/master/tests/TestCase/I18n/Parser/PoFileParserTest.php#L59-L102 So either remove them all or add them all!? However both ways will cause "noise".

Comment (dereuromark) [reply tests/TestCase/I18n/Parser/PoFileParserTest.php]: I guess in that case you can ignore my comment.

Comment (burzum) [reply tests/TestCase/I18n/Parser/PoFileParserTest.php]: We should agree on something and putt it into the coding standard? What do you think?

Comment (dereuromark) [reply tests/TestCase/I18n/Parser/PoFileParserTest.php]: Yeah, but can then be unrelated to this PR.

File: src/I18n/Translator.php (hunk 4 above)
Comment (markstory) [reply src/I18n/Translator.php]: No. I was just commenting on it for future us.

**Pull request merged**